cpp/include/cudf/table/row_operator3.cuh

```diff
@@ -0,0 +1,315 @@
+
+    column_device_view lcol = lhs;
+    column_device_view rcol = rhs;
+    while (lcol.type().id() == type_id::STRUCT) {
+      bool const lhs_is_null{lcol.is_null(lhs_element_index)};
+      bool const rhs_is_null{rcol.is_null(rhs_element_index)};
+
+      if (lhs_is_null or rhs_is_null) {  // atleast one is null
+        state           = null_compare(lhs_is_null, rhs_is_null, null_precedence);
+        last_null_depth = depth;
+        return thrust::make_pair(state, last_null_depth);
+      }
+
+      lcol = lcol.children()[0];
+      rcol = rcol.children()[0];
+      ++depth;
+    }
+
+    if (state == weak_ordering::EQUIVALENT) {
+      auto comparator = element_relational_comparator{nulls, lcol, rcol, null_precedence};
+      thrust::tie(state, last_null_depth) = cudf::type_dispatcher<non_nested_id_to_type>(
+        lcol.type(), comparator, lhs_element_index, rhs_element_index);
+    }
+
+    return thrust::make_pair(state, last_null_depth);
+  }
+
+ private:
+  column_device_view lhs;
+  column_device_view rhs;
+  Nullate nulls;
+  null_order null_precedence{};
+  int depth{std::numeric_limits<int>::max()};
+};
+
+/**
+ * @brief Computes whether one row is lexicographically *less* than another row.
+ *
+ * Lexicographic ordering is determined by:
+ * - Two rows are compared element by element.
+ * - The first mismatching element defines which row is lexicographically less
+ * or greater than the other.
+ *
+ * Lexicographic ordering is exactly equivalent to doing an alphabetical sort of
+ * two words, for example, `aac` would be *less* than (or precede) `abb`. The
+ * second letter in both words is the first non-equal letter, and `a < b`, thus
+ * `aac < abb`.
+ *
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <typename Nullate>
+class device_row_comparator {
+ public:
+  /**
+   * @brief Construct a function object for performing a lexicographic
+   * comparison between the rows of two tables.
+   *
+   * @throws cudf::logic_error if `lhs.num_columns() != rhs.num_columns()`
+   * @throws cudf::logic_error if column types of `lhs` and `rhs` are not comparable.
+   *
+   * @param lhs The first table
+   * @param rhs The second table (may be the same table as `lhs`)
+   * @param has_nulls Indicates if either input table contains columns with nulls.
+   * @param column_order Optional, device array the same length as a row that
+   * indicates the desired ascending/descending order of each column in a row.
+   * If `nullptr`, it is assumed all columns are sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row
+   * and indicates how null values compare to all other for every column. If
+   * it is nullptr, then null precedence would be `null_order::BEFORE` for all
+   * columns.
+   */
+  device_row_comparator(Nullate has_nulls,
+                        table_device_view lhs,
+                        table_device_view rhs,
```

Comment:
Something I've realized here is that this comparator can't be constructed from any old `table_device_view`. It has to be constructed from a `preprocessed_table`. So the constructor should limit that somehow via a type unique to `preprocessed_table`, or make this constructor private so only the owning comparator types can actually construct it.